tests/tests.cmake

```diff
@@ -170,11 +185,165 @@ function(add_xc7_test)
+
+        set(test_name "${name}-${board}")
+
+        set(device_channels_dir ${CMAKE_BINARY_DIR}/channels)
+        set(device_channels ${device_channels_dir}/${device}.db)
+        if(NOT TARGET ${device}-channels-db)
+            # Build the channels.db file required by fasm2bels
+            add_custom_command(
+                OUTPUT ${device_channels}
+                COMMAND
+                    ${CMAKE_COMMAND} -E make_directory ${device_channels_dir}
+                COMMAND
+                    python3 -mfasm2bels.database.create_channels
+                        --db-root ${PRJXRAY_DB_DIR}/${device_family}
+                        --part ${part}
+                        --connection-database ${device_channels}
+                )
+
+            add_custom_target(${device}-channels-db DEPENDS ${device_channels})
+        endif()
+
+        find_program(BITREAD bitread REQUIRED)
+
+        set(output_dir ${CMAKE_CURRENT_BINARY_DIR}/${board})
+        set(bit ${output_dir}/${name}.bit)
+        set(bit_target xc7-${test_name}-bit)
+
+        # Run fasm2bels to get logical and physical netlists
+        set(netlist ${output_dir}/${name}.bit.netlist)
+        set(phys ${output_dir}/${name}.bit.phys)
+        set(xdc ${output_dir}/${name}.bit.xdc)
+        set(fasm ${output_dir}/${name}.bit.fasm)
+        add_custom_command(
+            OUTPUT ${netlist} ${phys} ${xdc} ${fasm}
+            COMMAND
+                python3 -mfasm2bels
+                    --db_root ${PRJXRAY_DB_DIR}/${device_family}
+                    --part ${part}
+                    --connection_database ${device_channels}
+                    --bitread ${BITREAD}
+                    --bit_file ${bit}
+                    --fasm_file ${fasm}
+                    --logical_netlist ${netlist}
+                    --physical_netlist ${phys}
+                    --interchange_xdc ${xdc}
+                    --interchange_capnp_schema_dir ${INTERCHANGE_SCHEMA_PATH}
+            DEPENDS
+                xc7-${test_name}-bit
+                ${device}-channels-db
+        )
+
+        add_custom_target(xc7-${test_name}-fasm2bels DEPENDS ${netlist} ${phys} ${xdc} ${fasm})
+
+        # DCP generation target
+        set(dcp ${output_dir}/${name}.bit.dcp)
+        add_custom_command(
+            OUTPUT ${dcp}
+            COMMAND ${CMAKE_COMMAND} -E env
+                RAPIDWRIGHT_PATH=${RAPIDWRIGHT_PATH}
+                ${INVOKE_RAPIDWRIGHT}
+                com.xilinx.rapidwright.interchange.PhysicalNetlistToDcp
+                ${netlist} ${phys} ${xdc} ${dcp}
+            DEPENDS
+                ${INVOKE_RAPIDWRIGHT}
+                xc7-${test_name}-fasm2bels
+        )
+
+        add_custom_target(xc7-${test_name}-fasm2bels-dcp DEPENDS ${dcp})
+
+        # Generate tcl script to run DCP to bitstream generation
+        set(tcl ${output_dir}/runme.tcl)
+        set(vivado_bit ${output_dir}/${name}.vivado.bit)
+        add_custom_command(
```

Comment:
I'd rather write a TCL script that uses env vars and stores them in a file than crafting it from CMake using `echo`. You can then run `${CMAKE_COMMAND} -E env ...` to set them. Now it is kinda ok but I expect issues when switching to other OSes / shells in the future that may not recognize `echo`.

Reply:
Sure, makes sense. I have created a static runme.tcl with env variables.